src/tools/rbd/action/Create.cc

```diff
@@ -25,7 +26,174 @@ void get_arguments(po::options_description *positional,
                    po::options_description *options) {
   at::add_image_spec_options(positional, options, at::ARGUMENT_MODIFIER_NONE);
   at::add_create_image_options(options, true);
+  options->add_options()
+    (at::IMAGE_THICK_PROVISION.c_str(), po::bool_switch(), "fully allocate storage and zero image");
   at::add_size_option(options);
+  at::add_no_progress_option(options);
+}
+
+void thick_provision_writer_completion(rbd_completion_t, void *);
+
+struct thick_provision_writer {
+  librbd::Image *image;
+  Mutex lock;
+  Cond cond;
+  bufferlist bl;
+  const int chunk_size;
+  const int block_size;
+  uint64_t concurr;
+  struct {
+    uint64_t in_flight;
+    uint64_t complete_count;
+    int io_error;
+  } io_status;
+
+  // Constructor
+  explicit thick_provision_writer(librbd::Image *i)
+    : image(i),
+      lock("thick_provision_writer::lock"),
+      chunk_size(1024 * 1024), // 1MB
+      block_size(512) // 512 Bytes
+  {
+    bl.append_zero(block_size);
+    concurr = g_conf->get_val<int64_t>("rbd_concurrent_management_ops");
+    io_status.in_flight = 0;
+    io_status.complete_count = 0;
+    io_status.io_error = 0;
+  }
+
+  int start_io(uint64_t write_offset) {
+    Mutex::Locker l(lock);
+    io_status.in_flight++;
+    if (io_status.in_flight > concurr) {
+      return -EINVAL;
+    }
+
+    librbd::RBD::AioCompletion *c =
+      new librbd::RBD::AioCompletion(this, thick_provision_writer_completion);
+    int r = image->aio_writesame(write_offset, chunk_size, bl,
+                    c, LIBRADOS_OP_FLAG_FADVISE_SEQUENTIAL);
+    if (r < 0) {
+      io_status.io_error = r;
+    }
+    return r;
+  }
+
+  int wait_for(uint64_t max) {
+    Mutex::Locker l(lock);
+    int r = io_status.io_error;
+
+    while (io_status.in_flight > max) {
+      utime_t dur;
+      dur.set_from_double(.2);
+      cond.WaitInterval(lock, dur);
+    }
+    return r;
+  }
+};
+
+void thick_provision_writer_completion(rbd_completion_t rc, void *pc)
+{
+  librbd::RBD::AioCompletion *ac = (librbd::RBD::AioCompletion *)rc;
+  thick_provision_writer *tc = static_cast<thick_provision_writer *>(pc);
+
+  bool err = false;
+  int r = ac->get_return_value();
+  if (r != 0) {
+    err = true;
+  }
+  tc->lock.Lock();
+  if (err == true) {
+    tc->io_status.io_error = r;
+  }
+  tc->io_status.in_flight--;
+  tc->io_status.complete_count++;
+  tc->cond.Signal();
+  tc->lock.Unlock();
+  ac->release();
+}
+
+int write_data(librbd::Image &image, bool no_progress)
+{
+  uint64_t image_size;
+  int r = 0;
+  utils::ProgressContext pc("Thick provisioning", no_progress);
+
+  if (image.size(&image_size) != 0) {
+    return -EINVAL;
+  }
+
+  thick_provision_writer tpw(&image);
+  uint64_t chunk_count = (image_size + tpw.chunk_size - 1) / tpw.chunk_size;
+  uint64_t target_chunk = 0;
+  uint64_t write_offset_byte;
+  uint64_t complete_bytes;
+  int aios = 0;
+
+  for (;;) {
+    tpw.lock.Lock();
+    aios = tpw.concurr - tpw.io_status.in_flight;
```

Comment:
You shouldn't have to track this -- if you follow the example in ```Bench.cc```, you could ```tpw.wait_for(tpw.concurr - 1)``` and just schedule one AIO op, and allow the loop to bring you back here to see if you can immediately schedule another.